### ISSUE TITLE
Pd externals: use relative file paths

### ISSUE THREAD

**Issue by mgeier** [SoundScapeRenderer/ssr#265]
Currently, specifying a file (e.g. HRIR file) works only if an absolute path is given or if the path happens to be relative to Pd's working directory.

The path should, however, be interpreted as relative to the directory of the containing Pd patch (absolute paths should keep working).

Here's some code how Pd itself deals with relative paths: https://github.com/pure-data/pure-data/blob/eeef4ba9130d3182146927c37fa57d61bbff0f0b/src/s_path.c#L353-L408

However, this also opens the file, but we only need the path.

We could use our https://github.com/SoundScapeRenderer/ssr/blob/e50e06ea41917f2e38ecd1429a349c9dfaf703da/src/posixpathtools.h, but for Windows compatibility it would be better to rewrite this with https://en.cppreference.com/w/cpp/header/filesystem.

**Comment by umlaeute** [comment]
typically, you use `canvas_open()` to find a file in the patch's search paths, and on success close the returned filedescriptor with `sys_close()` (do not use `close()` if you don't need file handle leaks on Windows).

it's a bit clumsy but it works :-)

**Comment by mgeier** [comment]
Thanks @umlaeute, I tried this on https://github.com/AudioSceneDescriptionFormat/asdf-rust/pull/14, which is a different repo but has a similar problem.

I copy-pasted the re-combination of the path from the Pd sources, I hope that's correct.
It's indeed a bit clumsy.

I can implement the same thing here.

**Comment by umlaeute** [comment]
well, i would instead go for 

~~~C
char filename[MAXPDSTRING];
snprintf(filename, MAXPDSTRING-1, "%s/%s", dirresult, nameresult);
filename[MAXPDSTRING-1]=0;
~~~

which looks much more presentable...

**Comment by mgeier** [comment]
Wow, thanks, that's indeed much nicer!

I've added it to https://github.com/AudioSceneDescriptionFormat/asdf-rust/pull/14.

I'm obviously not very comfortable with the C formatting functions, but isn't the result already guaranteed to be zero-terminated?

At least that's the impression I get from https://en.cppreference.com/w/c/io/fprintf.

> At most `bufsz - 1` characters are written. The resulting character string will be terminated with a null character, unless `bufsz` is zero.

So I would expect that this works (assuming `MAXPDSTRING > 0`):

```c
char filename[MAXPDSTRING];
snprintf(filename, MAXPDSTRING, "%s/%s", dirresult, nameresult);
```

Or am I missing something?

BTW, you could also suggest your improvement to Pd proper, see

* https://github.com/pure-data/pure-data/blob/eeef4ba9130d3182146927c37fa57d61bbff0f0b/src/s_loader.c#L206-L211
* https://github.com/pure-data/pure-data/blob/eeef4ba9130d3182146927c37fa57d61bbff0f0b/src/s_loader.c#L178-L182
* https://github.com/pure-data/pure-data/blob/eeef4ba9130d3182146927c37fa57d61bbff0f0b/src/g_canvas.c#L204-L210

I'm not sure if the improvement can be applied in all cases.

**Comment by umlaeute** [comment]
> Or am I missing something?

probably not.
it's just that i always prefer to err on the safe side :-)

> BTW, you could also suggest your improvement to Pd proper, see

probably the current implementation is faster. i dunno.
Pd's code is full of `strcat`. some people [simply prefer it](https://stackoverflow.com/q/1772536/1169096)

**Comment by umlaeute** [comment]
> > Or am I missing something?
> 
> probably not.
> it's just that i always prefer to err on the safe side :-)

ah yes. and then there is [Windows](https://stackoverflow.com/a/13067917/1169096).

**Comment by mgeier** [comment]
Thanks for the infos @umlaeute! It's indeed quite unfortunate that MS has a similarly named function `_snprintf()` with a slightly different behavior.

But this doesn't keep me from using the standardized function according to its specification.

I've simplified the code in https://github.com/AudioSceneDescriptionFormat/asdf-rust/pull/14, but for the SSR externals I think I will use the helper functions from #268 instead. This way I'm not using any Pd-specific functions, in case somebody wants to try to compile the externals for Max (I have tried myself and failed, but maybe somebody else can do it).
